OpenRA.Mods.Common/UpdateRules/Rules/20221203/AbstractDocking.cs

```
@@ -0,0 +1,163 @@
+{
+	public class AbstractDocking : UpdateRule
+	{
+		readonly string[] refineyValues = { "DockAngle", "IsDragRequired", "DragOffset", "DragLength" };
+		readonly string[] harvesterValues = { "EnterCursor", "EnterBlockedCursor" };
+
+		readonly Dictionary<string, List<MiniYamlNode>> refineryNodes = new Dictionary<string, List<MiniYamlNode>>();
+		public override string Name => "Docking was abstracted from Refinery & Harvester.";
+
+		public override string Description =>
+			"Fields moved from Refinery to new trait DockHost, fields moved from Harvester to new trait DockClientManager and to DockHost";
+
+		public override IEnumerable<string> BeforeUpdateActors(ModData modData, List<MiniYamlNode> resolvedActors)
+		{
+			var harvesters = new Dictionary<string, HashSet<string>>();
+			var refineries = new List<string>();
+			foreach (var actorNode in resolvedActors)
+			{
+				var harvesterNode = actorNode.ChildrenMatching("Harvester", includeRemovals: false).FirstOrDefault();
+				if (harvesterNode != null)
+					harvesters[actorNode.Key] = harvesterNode.ChildrenMatching("DeliveryBuildings", includeRemovals: false).FirstOrDefault()?.NodeValue<HashSet<string>>() ?? new HashSet<string>();
+
+				var refineryNode = actorNode.ChildrenMatching("Refinery", includeRemovals: false).FirstOrDefault() ??
+					actorNode.ChildrenMatching("TiberianSunRefinery", includeRemovals: false).FirstOrDefault();
+				if (refineryNode != null)
+					refineries.Add(actorNode.Key);
+			}
+
+			foreach (var harvester in harvesters)
+			{
+				foreach (var deliveryBuilding in harvester.Value)
+				{
+					foreach (var refinery in refineries)
+					{
+						if (refinery == deliveryBuilding)
+						{
+							if (!refineryNodes.ContainsKey(refinery))
+								refineryNodes[refinery] = new List<MiniYamlNode>();
+
+							var node = new MiniYamlNode("DockType", deliveryBuilding.ToString());
+							if (!refineryNodes[refinery].Any(n => n.Key == node.Key))
```

I think it would be better to raise an exception than to check for duplicate entries here, since duplicate entries could have negative consequences in other parts of the codebase, we should not be encouraging them. Unless there is a valid reason for having duplicate entries that I'm missing here.

---

This is not checking for duplicates

---

Sorry I meant to say *permit* duplicates by only adding entries where they don't already exist. However now that I see these are UpdateRules it probably doesn't matter, since I presume it is only here for legacy compatibility reasons.